src/utils/spo.ts

```diff
@@ -1453,5 +1456,266 @@ export const spo = {
     const webProperties: WebProperties = await request.get<WebProperties>(requestOptions);
 
     return webProperties;
+  },
+
+  /**
+  * Gets file properties by server relative file url
+  * Returns a ListItemAllFields object
+  * @param webUrl Web url
+  * @param url The url of the list
+  * @param logger The logger object
+  * @param verbose If in verbose mode
+  */
+  async getFileAsListItemByUrl(webUrl: string, url: string, logger?: Logger, verbose?: boolean): Promise<any> {
+    if (verbose && logger) {
+      logger.logToStderr(`Getting the file properties with url ${url}`);
+    }
+
+    let requestUrl = `${webUrl}/_api/web/GetFileByServerRelativePath(DecodedUrl=@f)?$expand=ListItemAllFields`;
+    const serverRelativePath = urlUtil.getServerRelativePath(webUrl, url);
+    requestUrl += `&@f='${formatting.encodeQueryParameter(serverRelativePath)}'`;
+
+    const requestOptions: CliRequestOptions = {
+      url: requestUrl,
+      headers: {
+        'accept': 'application/json;odata=nometadata'
+      },
+
+      responseType: 'json'
+    };
+
+    const file = await request.get<any>(requestOptions);
+
+    return file.ListItemAllFields;
+  },
+
+  /**
+  * Gets a list items ID
+  * Returns the list id as string
+  * @param requestUrl The url which has to be used to make the request
+  * @param logger The logger object
+  * @param verbose If in verbose mode
+  */
+  async getListId(requestUrl: string, logger?: Logger, verbose?: boolean): Promise<string> {
+    if (verbose && logger) {
+      logger.logToStderr(`Getting list id...`);
+    }
+
+    const listRequestOptions: CliRequestOptions = {
+      url: `${requestUrl}?$select=Id`,
+      headers: {
+        'accept': 'application/json;odata=nometadata'
+      },
+      responseType: 'json'
+    };
+
+    const list = await request.get<{ Id: string; }>(listRequestOptions);
+    return list.Id;
+  },
+
+  /**
+  * Updates a list item with system update
+  * Returns the updated listitem object
+  * @param webUrl Web url
+  * @param requestUrl The url which has to be used to make the request
+  * @param id The id of the list item
+  * @param options An object of the properties that should be updated
+  * @param contentTypeName The name of the content type to update
+  * @param logger The logger object
+  * @param verbose If in verbose mode
+  */
+  async systemUpdateListItem(webUrl: string, requestUrl: string, id: string, options?: object, contentTypeName?: string, logger?: Logger, verbose?: boolean): Promise<any> {
```

Comment:
Hi @waldekmastykarz,

This `spo` function is called directly or indirectly from various commands. Depending on which command this function is called, the `requestUrl` parameter is different. below an overview:

| Command | RequestUrl | 
|---------|------------|
| spo page add | https://contoso.sharepoint.com/_api/web/GetList('%2Fsitepages') |
| spo page set| https://contoso.sharepoint.com/_api/web/GetList('%2Fsitepages') |
| spo listitem set (with `listTitle` option) | https://contoso.sharepoint.com/_api/web/lists/getByTitle('LISTTITLE') |
| spo listitem set (with `listUrl` option) | https://contoso.sharepoint.com/_api/web/GetList('LISTURL') |
| spo listitem set (with `listId` option) | https://contoso.sharepoint.com/_api/web/lists(guid'LISTID') |

What do you suggest to make this function simpler?

Looking forward to hear your opion

Reply:
Looking at all these URLs, they're all the same, except for the API suffix/method. Since the `_api` fragment denotes the end of the web URL and the begin of the API reference, what if we removed the `webUrl` and extracted the webUrl from the `requestUrl`? That way, we don't require the consumer to pass the same information twice.